src/process/unix/SDL_process.c

```diff
@@ -0,0 +1,314 @@
+{
+    // Keep the malloc() before exec() so that an OOM won't run a process at all
+    SDL_Process *process = SDL_malloc(sizeof(SDL_Process));
+
+    if (!process) {
+        return NULL;
+    }
+
+    process->flags = flags;
+
+    if ((flags & SDL_PROCESS_STDIN) && (pipe(process->stdin_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDOUT) && (pipe(process->stdout_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDERR) && (pipe(process->stderr_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    process->pid = fork();
+
+    if (process->pid < 0) {
+        SDL_SetError("Could not fork(): %s", strerror(errno));
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            close(process->stderr_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+
+        return NULL;
+    } else if (process->pid == 0) {
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[WRITE_END]);
+            dup2(process->stdin_pipe[READ_END], STDIN_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            dup2(process->stdout_pipe[WRITE_END], STDOUT_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            dup2(process->stderr_pipe[WRITE_END], STDERR_FILENO);
+        }
+
+        char **mutable_args = dupstrlist(args);
+        char **mutable_env = env ? dupstrlist(env) : NULL;
+
+        // We are in a new process; don't bother freeing because either exec*() or exit() will succeed
+
+        if (!mutable_args || (env && !mutable_env)) {
+            fprintf(stderr, "Could not clone args/env: %s\n", SDL_GetError());
+            exit(1);
```

Comment:
`exit()` is not async-signal-safe (because it calls arbitrary `atexit()` handlers). Always use `_exit()` in a forked child process.